File: python/atlasent/verticals/contract_actions.py
"""Contract action protect wrappers — Phase 4 vertical.

Covers contract execution (irreversible, critical) and amendment
(partial reversibility, high risk). Both actions require human review
and are fail-closed.
"""

from __future__ import annotations

from typing import Any, Literal

from atlasent.authorize import protect
from atlasent.models import Permit

ContractActionType = Literal[
    "contract.execute",
    "contract.amend",
]

_CONTRACT_RISK: dict[str, str] = {
    "contract.execute": "critical",
    "contract.amend": "high",
}

_CONTRACT_REVERSIBILITY: dict[str, str] = {
    "contract.execute": "irreversible",
    "contract.amend": "partial",
}

_MACHINE_EXECUTABLE = False
_FAIL_CLOSED = True


def protect_contract_action(
    action: ContractActionType,
    contract_id: str,
    authorized_by: str,
    **kwargs: Any,
) -> Permit:
    """Generic contract action protect wrapper.

    Args:
        action: One of the ``ContractActionType`` literals.
        contract_id: The contract being acted upon.
        authorized_by: Agent or human ID authorising the action.
        **kwargs: Action-specific fields. For ``contract.amend``,
            ``amendment_description`` is required. Optional overrides:
            ``assigned_to_role``, ``wait_ms``.

    Raises:
        ValueError: If ``amendment_description`` is missing for
            ``contract.amend``.
    """
    if action == "contract.amend" and not kwargs.get("amendment_description"):
        raise ValueError(
            "Contract action 'contract.amend' requires 'amendment_description'"
        )

    risk_level = _CONTRACT_RISK.get(action, "high")
    reversibility = _CONTRACT_REVERSIBILITY.get(action, "partial")
    quorum = "simple_majority" if risk_level == "critical" else "single_approver"

    context: dict[str, Any] = {
        "machine_executable": _MACHINE_EXECUTABLE,
        "risk_level": risk_level,
        "fail_closed": _FAIL_CLOSED,
        "contract_id": contract_id,
        "contract_action": action,
        "reversibility": reversibility,
        "hitl_escalation": {
            "assigned_to_role": kwargs.get("assigned_to_role", "legal-approver"),
            "quorum_required": kwargs.get("quorum_required", quorum),
            "wait_ms": kwargs.get("wait_ms", 86_400_000),  # 24 h
        },
    }

    if kwargs.get("amendment_description"):
        context["amendment_description"] = kwargs["amendment_description"]

    return protect(agent=authorized_by, action=action, context=context)


def protect_contract_execution(
    contract_id: str,
    authorized_by: str,
    **kwargs: Any,
) -> Permit:
    """Convenience wrapper for ``contract.execute``.

    Args:
        contract_id: The contract being executed.
        authorized_by: Authorising agent/human ID.
        **kwargs: Optional overrides forwarded to ``protect_contract_action``.
    """
    return protect_contract_action(
        "contract.execute",
        contract_id,
        authorized_by,
        **kwargs,
    )

File: typescript/src/verticals/contractActions.ts
import { protectOrEscalate } from "../approvalRuntime.js";
import { buildActionContext, flattenActionContext } from "../actionContext.js";
import type { ApprovalPermit } from "../approvalRuntime.js";
import type { EscalationHandle } from "../approvalRuntime.js";

export type ContractActionType =
  | "contract.execute"
  | "contract.amend";

const CONTRACT_ACTION_RISK: Record<ContractActionType, "critical" | "high"> = {
  "contract.execute": "critical",
  "contract.amend": "high",
};

export interface ContractActionOptions {
  action: ContractActionType;
  contractId: string;
  authorizedBy: string;
  counterparty: string;
  legalReviewId?: string;
  estimatedValue?: number;
  currency?: string;
  effectiveDate?: string;
  amendmentDescription?: string;
  assignedToRole?: string;
  waitMs?: number;
  onEscalationCreated?: (handle: EscalationHandle) => void;
  apiKey?: string;
  baseUrl?: string;
}

export async function protectContractAction(
  opts: ContractActionOptions,
): Promise<ApprovalPermit> {
  if (opts.action === "contract.amend" && !opts.amendmentDescription) {
    throw new TypeError(
      `Contract action 'contract.amend' requires 'amendmentDescription'`,
    );
  }

  const riskLevel = CONTRACT_ACTION_RISK[opts.action] ?? "high";

  const ctx = buildActionContext({
    actor: {
      id: opts.authorizedBy,
      type: "human",
    },
    resource: {
      id: opts.contractId,
      type: "contract",
      sensitivity: "restricted",
    },
    environment: "production",
    action_meta: {
      risk_level: riskLevel,
      reversibility: opts.action === "contract.execute" ? "irreversible" : "partial",
      description: `Contract action '${opts.action}' on contract '${opts.contractId}' with counterparty '${opts.counterparty}'`,
      ...(opts.estimatedValue !== undefined ? { estimated_amount: opts.estimatedValue } : {}),
      ...(opts.currency !== undefined ? { currency: opts.currency } : {}),
    },
    extra: {
      contract_action: opts.action,
      machine_executable: false,
      fail_closed: true,
      counterparty: opts.counterparty,
      ...(opts.legalReviewId !== undefined ? { legal_review_id: opts.legalReviewId } : {}),
      ...(opts.effectiveDate !== undefined ? { effective_date: opts.effectiveDate } : {}),
      ...(opts.amendmentDescription !== undefined ? { amendment_description: opts.amendmentDescription } : {}),
    },
  });

  return protectOrEscalate(
    {
      action: opts.action,
      agent: opts.authorizedBy,
      context: flattenActionContext(ctx),
    },
    {
      escalationReason: `Contract action '${opts.action}' on '${opts.contractId}' requires legal review — machine_executable: false`,
      assignedToRole: opts.assignedToRole ?? (riskLevel === "critical" ? "legal-director" : "legal-reviewer"),
      quorumRequired: riskLevel === "critical" ? "simple_majority" : "single_approver",
      waitMs: opts.waitMs ?? 48 * 60 * 60 * 1000,
      ...(opts.onEscalationCreated !== undefined ? { onEscalationCreated: opts.onEscalationCreated } : {}),
      ...(opts.apiKey !== undefined ? { apiKey: opts.apiKey } : {}),
      ...(opts.baseUrl !== undefined ? { baseUrl: opts.baseUrl } : {}),
    },
  );
}

export async function protectContractExecution(
  opts: Omit<ContractActionOptions, "action">,
): Promise<ApprovalPermit> {
  return protectContractAction({ ...opts, action: "contract.execute" });
}
